shared/lib_shared_inverter.cpp

```diff
@@ -8,9 +9,93 @@ SharedInverter::SharedInverter(int inverterType, int numberOfInverters,
 	m_numInverters = numberOfInverters;
 	m_sandiaInverter = sandiaInverter;
 	m_partloadInverter = partloadInverter;
+	m_tempEnabled = false;
 }
 
-void SharedInverter::calculateACPower(const double powerDC_Watts, const double DCStringVoltage)
+bool sortByVoltage(const double i[], const double j[])
+{
+	return (i[0] < j[0]);
+}
+
+bool SharedInverter::setTempDerateCurves(double* curve1, double* curve2, double* curve3)
+{
+	if (curve1[0] < 0 || curve1[1] <= -99 || curve1[2] >= 0) return false;
+
+	// if multiple curves provided, partition operating V range
+	std::vector<double*> derateCurves;
+	derateCurves.push_back(curve1);
+	if (curve2 != NULL){
+		if (curve2[1] <= -99 || curve2[2] >= 0) return false;
+		derateCurves.push_back(curve2);
+	}
+	if (curve3 != NULL){
+		if (curve3[1] <= -99 || curve3[2] >= 0) return false;
+		derateCurves.push_back(curve3);
+	}
+
+	std::sort(derateCurves.begin(), derateCurves.end(), sortByVoltage);
+	m_tempStartC[0] = derateCurves[0][1];
+	m_tempSlope[0] = derateCurves[0][2];
+
+	if (derateCurves.size() > 1){
+		m_tempV[0] = (derateCurves[1][0] + derateCurves[0][0])/2;
+		m_tempStartC[1] = derateCurves[1][1];
+		m_tempSlope[1] = derateCurves[1][2];
+		if (derateCurves.size() > 2) {
+			m_tempV[1] = (derateCurves[2][0] + derateCurves[1][0])/2;
+			m_tempStartC[2] = derateCurves[2][1];
+			m_tempSlope[2] = derateCurves[2][2];
+		}
+	}
+
+	m_tempEnabled = true;
+	return true;
+}
+
+void SharedInverter::getTempDerateCurves(double* vParts, double* startC, double* slope) {
+	vParts[0] = m_tempV[0];
+	vParts[1] = m_tempV[1];
+	startC[0] = m_tempStartC[0];
+	startC[1] = m_tempStartC[1];
+	startC[2] = m_tempStartC[2];
+	slope[0] = m_tempSlope[0];
+	slope[1] = m_tempSlope[1];
+	slope[2] = m_tempSlope[2];
+}
+
+void SharedInverter::calculateTempDerate(double V, double T, double& pAC, double& eff, double& loss)
+{
+	if (eff == 0. || pAC == 0.) return;
+
+	double slope = 0.0;
+	double deltaT = 0.0;
+
+	if (m_tempV[0] == 0 || (m_tempV[0] != 0 && V <= m_tempV[0])){

```

Comment:
Why checking for 0 here, and again in line 79? What if there's only one curve specified? This seems to assume all three curves exist. What if there's only one curve, but the user inputs V > m_tempV[0]? What happens then? One possible logic fix would be to work backwards from the biggest existing curve until you find one with m_tempV[i] < V, otherwise use the first curve. 

Also totally possible I'm completely misunderstanding what's happening here. :)